torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -353,119 +353,137 @@ std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
-  int64_t mType = static_cast<int>(rpcMessage.type());
-  int64_t mId = rpcMessage.id();
-  memcpy((void*)tpMessage.metadata.data(), &mType, sizeof(int64_t));
-  memcpy(
-      (void*)(tpMessage.metadata.data() + sizeof(int64_t)),
-      &mId,
-      sizeof(int64_t));
+  buffers.payload = std::move(rpcMessage).movePayload();
+  tpMessage.payloads.push_back(tensorpipe::Message::Payload{
+      const_cast<char*>(buffers.payload.data()), buffers.payload.size()});
 
   // Tensors
-  tpMessage.tensors.reserve(tensors.size());
-  for (at::Tensor tensor : tensors) {
-    // Keep original user tensors and cloned sparse tensors
-    reservedTensors.push_back(tensor);
-    tensorpipe::Message::Tensor tpTensor;
-    torch::jit::Pickler pickler([&](const void* buf, size_t sz) -> size_t {
-      tpTensor.metadata.append(static_cast<const char*>(buf), sz);
-      return sz;
-    });
-    pickler.protocol();
-    pickler.pushIValue(tensor);
-    pickler.stop();
-    const auto& tensorDataVec = pickler.tensorData();
-    TORCH_INTERNAL_ASSERT(
-        tensorDataVec.size() == 1, "There should be single pickled tensor");
-    const auto& tensorData = tensorDataVec.front();
+  buffers.tensors = cloneSparseTensors(rpcMessage.tensors()).vec();
+  torch::jit::Pickler pickler([&](const void* buf, size_t sz) -> size_t {
+    buffers.pickle.insert(
+        buffers.pickle.end(),
+        static_cast<const char*>(buf),
+        static_cast<const char*>(buf) + sz);
+    return sz;
+  });
+  pickler.protocol();
+  pickler.pushIValue(buffers.tensors);
+  pickler.stop();
+  tpMessage.payloads.push_back(tensorpipe::Message::Payload{
+      buffers.pickle.data(), buffers.pickle.size()});
+  for (const auto& tensorData : pickler.tensorData()) {
     // Enforce memory copy if tensor is created from torch::from_blob, means
     // that the tensor doesn't own the memory.
     if (!tensorData.storageHasDeleter()) {
-      std::vector<uint8_t> storageData(tensorData.sizeInBytes());
-      memcpy(storageData.data(), tensorData.data(), storageData.size());
-      tpTensor.data = storageData.data();
-      copiedTensors.push_back(std::move(storageData));
+      std::vector<char> storageData(
+          tensorData.data(), tensorData.data() + tensorData.sizeInBytes());
+      tpMessage.tensors.push_back(
+          tensorpipe::Message::Tensor{storageData.data(), storageData.size()});
+      buffers.copiedTensors.push_back(std::move(storageData));
     } else {
-      tpTensor.data = (uint8_t*)(tensorData.data());
+      tpMessage.tensors.push_back(tensorpipe::Message::Tensor{
+          const_cast<char*>(tensorData.data()), tensorData.sizeInBytes()});
     }
-    tpTensor.length = tensorData.sizeInBytes();
-    tpMessage.tensors.push_back(std::move(tpTensor));
   }
 
-  return TensorPipeEntry{std::move(tpMessage),
-                         std::move(reservedTensors),
-                         std::move(copiedTensors)};
+  return std::make_tuple(std::move(tpMessage), std::move(buffers));
```

Comment:
What is the reason for explicitly returning `buffers`? Is the returned `buffers` used anywhere besides making sure the lifetime of payload (etc.) is correct?

Reply:
The buffers need to be returned so that the agent can keep it alive until the write has completed. There is no other purpose. See comment above for more discussion.